Source/Extensions/Blazorise.TreeView/TreeView.razor.cs

```diff
@@ -11,21 +17,130 @@ public partial class TreeView<TNode> : BaseComponent
 {
     #region Members
 
-    private TreeViewState<TNode> state = new()
+    private _TreeViewNode<TNode> treeViewNodeRef;
+
+    private TreeViewState<TNode> treeViewState = new()
     {
+        SelectedNodes = new List<TNode>(),
+        SelectionMode = TreeViewSelectionMode.Single,
     };
 
+    private List<TreeViewNodeState<TNode>> treeViewNodeStates;
+
     #endregion
 
     #region Methods
 
+    public override async Task SetParametersAsync( ParameterView parameters )
+    {
+        bool nodesChanged = parameters.TryGetValue<IEnumerable<TNode>>( nameof( Nodes ), out var paramNodes ) && !paramNodes.AreEqual( Nodes );
+        bool selectedNodeChanged = parameters.TryGetValue<TNode>( nameof( SelectedNode ), out var paramSelectedNode ) && !paramSelectedNode.IsEqual( SelectedNode );
+        bool selectedNodesChanged = parameters.TryGetValue<IList<TNode>>( nameof( SelectedNodes ), out var paramSelectedNodes ) && !paramSelectedNodes.AreEqual( SelectedNodes );
+
+        if ( selectedNodeChanged )
+        {
+            treeViewState = treeViewState with { SelectedNode = paramSelectedNode };
+        }
+
+        if ( selectedNodesChanged )
+        {
+            treeViewState = treeViewState with { SelectedNodes = paramSelectedNodes };
+        }
+
+        await base.SetParametersAsync( parameters );
+
+        if ( nodesChanged )
+        {
+            treeViewNodeStates = new();
+
+            await foreach ( var nodeState in paramNodes.ToNodeStates( HasChildNodesAsync, HasChildNodes, ExpandedNodes.Intersect( paramNodes ?? Enumerable.Empty<TNode>() ).Any() ) )
+            {
+                treeViewNodeStates.Add( nodeState );
+            }
+        }
+
+        if ( selectedNodeChanged )
+        {
+            await SelectedNodeChanged.InvokeAsync( treeViewState.SelectedNode );
+        }
+
+        if ( selectedNodesChanged )
+        {
+            await SelectedNodesChanged.InvokeAsync( treeViewState.SelectedNodes );
+        }

```

Comment:
If we are on `SetParametersAsync`, and either of these have been changed then it means the Parameter was already changed / mutated and there is no need to invoke these? We didn't really do any logic to mutate anything did we?

Reply:
That's not my point. My point is that we don't need to call the Changed at all for either. Because at this time, the change has already been done and the consumer already knows about it.

Reply:
🤔 I guess you're right.

Reply:
We need these. If the user updates `SelectedNode` though the code and through the parameter, maybe he also listens for changes with `SelectedNodeChanged`. If we don't do this, then the user will never get the notification.

Reply:
Well exactly. He updated he doesn't need notification, it's redundant I believe? The notification is there due to the declarative nature of blazor. If you go to a regular Blazor input, a text for example and programmatically set text to something does it notify changed back? I doubt so, would be interesting to test. 

I believe this will trigger an unnecessary re-render that's why I was advocating to not having it. But if you feel we should still treat as notification worthy.

Reply:
I still think notification must be invoked.

Reply:
I admit the defeat

```razor
<div>
    <Button Color="Color.Primary" Clicked="@(()=>Description = "TEST")">TEST</Button>
</div>
<div>
    <input type="text" value="@Description" @oninput="@OnInput" />
</div>
<div>
    Value : @Description
</div>
@code {
    string Description;

    Task OnInput( ChangeEventArgs eventArgs )
    {
        Description = eventArgs?.Value?.ToString();
        Console.WriteLine( $"onchanged: {Description}" );
        return Task.CompletedTask;
    }
}
```

Reply:
haha noone's defeating anyone, this ain't a fight 🗡️🗡️ 

I actually meant for us to test the Native blazor input component, that'd be a closer comparisition I believe. But I believe the result would be the same.

Reply:
> I actually meant for us to test the Native blazor input component

Can you try it?

Reply:
OK I have tried it. The event is not called after the binded value is changed.

```razor
<div>
    <Button Color="Color.Primary" Clicked="@(()=>exampleModel.Name = "TEST")">TEST</Button>
</div>
<div>
    <EditForm Model="@exampleModel">
        <DataAnnotationsValidator />

        <MyInputText id="name" Value="@exampleModel.Name" ValueChanged="@OnInput" ValueExpression="@(()=>exampleModel.Name)" />
    </EditForm>

</div>
<div>
    Value : @exampleModel.Name
</div>

@code {
    public class ExampleModel
    {
        [Required]
        [StringLength( 10, ErrorMessage = "Name is too long." )]
        public string? Name { get; set; }
    }
    private ExampleModel exampleModel = new();
    Task OnInput( string value )
    {
        exampleModel.Name = value;
        Console.WriteLine( $"onchanged: {exampleModel.Name}" );
        return Task.CompletedTask;
    }
}
```

and

**MyInputText.razor**

```razor
@inherits Microsoft.AspNetCore.Components.Forms.InputText
<input @attributes="@AdditionalAttributes" class="@CssClass" @bind="@CurrentValueAsString" @bind:event="oninput" />
```